pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/QueryRunner.java

```diff
@@ -143,22 +148,51 @@ public void shutDown()
   }
 
   public void processQuery(DistributedStagePlan distributedStagePlan, Map<String, String> requestMetadataMap) {
+    try {
     long requestId = Long.parseLong(requestMetadataMap.get(QueryConfig.KEY_OF_BROKER_REQUEST_ID));
     long timeoutMs = Long.parseLong(requestMetadataMap.get(QueryConfig.KEY_OF_BROKER_REQUEST_TIMEOUT_MS));
     boolean isTraceEnabled =
         Boolean.parseBoolean(requestMetadataMap.getOrDefault(CommonConstants.Broker.Request.TRACE, "false"));
     long deadlineMs = System.currentTimeMillis() + timeoutMs;
+    // run OpChain
     if (isLeafStage(distributedStagePlan)) {
-      OpChain rootOperator = compileLeafStage(distributedStagePlan, requestMetadataMap, timeoutMs, deadlineMs,
-          requestId);
+      // pre-stage execution for all pipeline breakers
+      // TODO: pipeline breaker is now only supported by leaf stage, to be supported by all OpChain
+      PipelineBreakerContext pipelineBreakerContext = executePipelineBreakers(_intermScheduler, distributedStagePlan,
+          timeoutMs, deadlineMs, requestId, isTraceEnabled);
```

Comment:
TODO: this execution is done on a dedicated threadpool, but is awaiting on the GRPC runner main thread (parked)
we should come up with a better wait and continue model

Reply:
this is addressed via #10791